datafusion/physical-plan/src/joins/hash_join.rs

```diff
@@ -1210,6 +1212,16 @@ fn eq_dyn_null(
     right: &dyn Array,
     null_equals_null: bool,
 ) -> Result<BooleanArray, ArrowError> {
+    // Nested datatypes cannot use the underlying not_distinct function and must use a special
+    // implementation
+    // <https://github.com/apache/datafusion/issues/10749>
```

Comment:
I think the issue is that the underlying `eq` kernel doesn't work for nested structs

I filed https://github.com/apache/arrow-rs/issues/5960 to track this upstream in arrow-rs


```suggestion
    // eq does not yet support nested structs.
    // https://github.com/apache/arrow-rs/issues/5960 
```

Reply:
Actually, after discussions with @jayzhan211  I don't think arrow-rs will ever support nested structs